Avoid changing current cudnn benchmark (#12020)
PyTorchLightning/pytorch-lightning: docs/source/common/trainer.rst

|

The value (``True`` or ``False``) to set ``torch.backends.cudnn.benchmark`` to. If not specified, the value

> set in the current session will be unchanged.

Reviewer: ```suggestion
set in the current session will be used.
```

> Setting this flag to ``True`` is likely to increase the speed of your system if your input sizes don't

Reviewer: ```suggestion
Setting this flag to ``True`` is likely going to increase the speed of your system if your input sizes don't
```

change. However, if it does, then it will likely make your system slower.

The speedup comes from allowing the cudnn auto-tuner to find the best
algorithm for the hardware `[see discussion here]
<https://discuss.pytorch.org/t/what-does-torch-backends-cudnn-benchmark-do/5936>`_.

Example::

>     # default used by the Trainer (current setting for torch.backends.cudnn.benchmark is unchanged)

Reviewer 2: minor nit for docs:

```suggestion
# default used by the Trainer (current setting for ``torch.backends.cudnn.benchmark`` is unchanged)
```

Author: I think literal blocks have to be indented in RST https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html

Author: Actually, since this comment is also in the literal example block, it is code also. So no backticks are needed if we are to follow the convention in the rest of this file

Reviewer 2: Apologies for the indentation suggestion, it was by fault.

And sounds good, you are right!

>     # default used by the Trainer (current setting for torch.backends.cudnn.benchmark is unchanged)

Reviewer: ```suggestion
    # default used by the Trainer (will use whatever the current value for torch.backends.cudnn.benchmark is)
```

:paramref:`~pytorch_lightning.trainer.Trainer.deterministic` are set, the value for
``torch.backends.cudnn.benchmark`` set in the current session will be used. If
:paramref:`~pytorch_lightning.trainer.Trainer.deterministic` is set to ``True``, this will default to ``False``.
You can read more about the interaction of ``torch.backends.cudnn.benchmark`` and ``torch.backends.cudnn.deterministic``
`here <https://pytorch.org/docs/stable/notes/randomness.html#cuda-convolution-benchmarking>`__

Setting this flag to ``True`` is likely to increase the speed of your system if your input sizes don't
change. However, if they do, then it might make your system slower. The CUDNN auto-tuner will try to find the best
algorithm for the hardware when a new input size is encountered. Read more about it
`here <https://discuss.pytorch.org/t/what-does-torch-backends-cudnn-benchmark-do/5936>`__.

Example::

    # default used by the Trainer (will use whatever the current value for torch.backends.cudnn.benchmark is)
    trainer = Trainer(benchmark=None)

    # you can overwrite the value
    trainer = Trainer(benchmark=False)

>     # defaults to False when deterministic is True

Reviewer 3: ```suggestion
    # `benchmark` defaults to False when deterministic is True
```
